### ISSUE TITLE
Can not install libyang -  fatal error: libyang/libyang.h: No such file or directory

### ISSUE THREAD

**Issue by obernat** [CESNET/libyang-python#19]
I'm trying to install libyang on Ubuntu 18.04.4 LTS, Python version 3.6.9, but I ends up with the following error:

admin@pc:[~]$ pip3 install libyang
Collecting libyang
  Downloading https://files.pythonhosted.org/packages/20/23/0b2445f42452f548401ae90f434d5e531e15a131f42570467fc722906c1e/libyang-1.6.1.tar.gz
Requirement already satisfied: cffi in /usr/local/lib/python3.6/dist-packages (from libyang)
Requirement already satisfied: pycparser in /usr/local/lib/python3.6/dist-packages (from cffi->libyang)
Installing collected packages: libyang
  Running setup.py install for libyang ... error
    Complete output from command /usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-wft4zgc_/libyang/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-8ybf8nd2-record/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.6
    creating build/lib.linux-x86_64-3.6/libyang
    copying libyang/data.py -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/keyed_list.py -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/context.py -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/diff.py -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/__init__.py -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/schema.py -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/xpath.py -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/util.py -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/log.py -> build/lib.linux-x86_64-3.6/libyang
    running egg_info
    writing libyang.egg-info/PKG-INFO
    writing dependency_links to libyang.egg-info/dependency_links.txt
    writing requirements to libyang.egg-info/requires.txt
    writing top-level names to libyang.egg-info/top_level.txt
    reading manifest file 'libyang.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'libyang.egg-info/SOURCES.txt'
    copying libyang/VERSION -> build/lib.linux-x86_64-3.6/libyang
    copying libyang/py.typed -> build/lib.linux-x86_64-3.6/libyang
    running build_ext
    generating cffi module 'build/temp.linux-x86_64-3.6/_libyang.c'
    creating build/temp.linux-x86_64-3.6
    building '_libyang' extension
    creating build/temp.linux-x86_64-3.6/build
    creating build/temp.linux-x86_64-3.6/build/temp.linux-x86_64-3.6
    x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.6m -c build/temp.linux-x86_64-3.6/_libyang.c -o build/temp.linux-x86_64-3.6/build/temp.linux-x86_64-3.6/_libyang.o -Werror -std=c99
    build/temp.linux-x86_64-3.6/_libyang.c:545:10: fatal error: libyang/libyang.h: No such file or directory
     #include <libyang/libyang.h>
              ^~~~~~~~~~~~~~~~~~~
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    
    ----------------------------------------
Command "/usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-wft4zgc_/libyang/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-8ybf8nd2-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-wft4zgc_/libyang/

I tried this on another machine with Centos 6 but the result was same.

**Comment by rjarry** [comment]
Hi, as described in the readme, you need to have libyang headers available when installing the python bindings.

https://github.com/CESNET/libyang-python#installation

If you need help about libyang installation itself, this should be a good starting point:

https://github.com/CESNET/libyang#building

**Comment by obernat** [comment]
Hello, thank you for quick response. 
I'm developing python library and I would like to use libyang as a dependency to my python package. Do I understand you correctly that there is no solution in a way that I define libyang as a dependency in pip so I get out of box libyang with python binding?

**Comment by rjarry** [comment]
You can build and install libyang into a python virtualenv. And then provide options to make sure that the python bindings find the library and headers. This also requires to force `-rpath` so that `libyang.so` is found by the dynamic loader at runtime. [I use this method for unit tests](https://github.com/CESNET/libyang-python/blob/master/tox-install.sh).

TL;DR

```bash
python3 -m venv venv
PREFIX=$PWD/venv
. $PREFIX/bin/activate
git clone https://github.com/CESNET/libyang $PREFIX/src/libyang
mkdir -p $PREFIX/build/libyang
cmake -DCMAKE_INSTALL_PREFIX="$PREFIX" -DCMAKE_INSTALL_LIBDIR=lib -H$PREFIX/src/libyang -B$PREFIX/build/libyang
make -C $PREFIX/build/libyang install
LIBYANG_HEADERS="$PREFIX/include" \
LIBYANG_LIBRARIES="$PREFIX/lib" \
LIBYANG_EXTRA_LDFLAGS="-Wl,--enable-new-dtags,-rpath=$PREFIX/lib" \
	python -m pip install libyang
```

**Comment by rjarry** [comment]
Mind that this method is only suitable for development and local tests.

For production, I advise to install libyang into a standard location.

**Comment by obernat** [comment]
I see, thank you. Thing is that I would like to be able to ship python package with libyang as a requirement so I would be able to tell "in this version of mySuperPythonLibrary the requirement is libyang 1.0.3" and this requirement could be fully satisfied via pip. Is there any plan for the future to make libyang part of pip package?
Thanks!

**Comment by rjarry** [comment]
Actually, this *was* supported until commits https://github.com/CESNET/libyang-python/commit/fae31204b91f3d14ca1b9c85686016f5a4367065 and https://github.com/CESNET/libyang-python/commit/7823e6af464efd9211c28df79c7a4ba9dcb4c7bd.

I removed that "feature" since it was a pain to maintain and also because it breaks using dynamic libyang plugins. Additionally, it made interaction with [sysrepo](https://github.com/sysrepo/sysrepo-python) a lot more complex since `libsysrepo.so` needs libyang as well.

**Comment by obernat** [comment]
Oh, I see. Thank you very much for replies, I think we can close this one.

**Comment by TribuneX** [comment]
I tried to replicate the steps posted by @rjarry for my own setup, but I can't get this to work on my M1 Pro with the last [libyang v1 release](https://github.com/CESNET/libyang/releases/tag/v1.0.240). 

I'am stuck at the following issue:
```
venv ❯ make -C $PREFIX/build/libyang install
[  1%] Building C object CMakeFiles/compat.dir/compat/compat.c.o
/Users/xxx/src/xxx/venv/src/libyang/compat/compat.c:64:13: error: implicitly declaring library function 'vsnprintf' with type 'int (char *, unsigned long, const char *, __builtin_va_list)' [-Werror,-Wimplicit-function-declaration]
    int l = vsnprintf(0, 0, fmt, ap2);
            ^
/Users/xxx/src/xxx/venv/src/libyang/compat/compat.c:64:13: note: include the header <stdio.h> or explicitly provide a declaration for 'vsnprintf'
/Users/xxx/src/xxx/venv/src/libyang/compat/compat.c:145:18: error: implicitly declaring library function 'strdup' with type 'char *(const char *)' [-Werror,-Wimplicit-function-declaration]
        retval = strdup(tmp);
                 ^
/Users/xxx/src/xxx/venv/src/libyang/compat/compat.c:145:18: note: include the header <string.h> or explicitly provide a declaration for 'strdup'
2 errors generated.
make[2]: *** [CMakeFiles/compat.dir/compat/compat.c.o] Error 1
make[1]: *** [CMakeFiles/compat.dir/all] Error 2
make: *** [all] Error 2
```
I tried to use llvm as described here: https://github.com/CESNET/libyang/issues/1459, but it does not support the required options for:
```
LIBYANG_HEADERS="$PREFIX/include" \
LIBYANG_LIBRARIES="$PREFIX/lib" \
LIBYANG_EXTRA_LDFLAGS="-Wl,--enable-new-dtags,-rpath=$PREFIX/lib" \
	python -m pip install libyang
```

Any hint how to get this working?

**Comment by rjarry** [comment]
Hi, 

this issue is closed and is not related to your problem.

I don't know if macos is a supported target of libyang. In any case you should sort this out with libyang maintainers first, before trying to use the python bindings :)

I don't know if llvm is supported by CFFI python extensions.
